<p>ProTN is an integrative pipeline that analyze DDA proteomics data
	obtained from MS. It perform a complete analysis of the raw files from
	<a href="https://www.thermofisher.com/it/en/home/industrial/mass-spectrometry/liquid-chromatography-mass-spectrometry-lc-ms/lc-ms-software/multi-omics-data-analysis/proteome-discoverer-software.html">Proteome
	Discoverer</a> (PD) or <a href="https://maxquant.org/">MaxQuant</a>
	(MQ), with their biological interpretation with enrichement and network
	analysis. ProTN executes a dual level analysis, at protein and peptide
	level.
</p>
<div id="workflow-of-protn" class="section level2">
  <ul id="collapse_ul" class="collapsible">
    <li id="collapse_li" class="active">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
    	    <h2>1. Workflow of ProTN</h2>
    	    <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
      	<p>Later in this section we describes in details the workflow of the
      		pipeline. In the first phases it imputes and normalizes the data
      		obtaining the matrix that is used in the enrichment analysis of the
      		proteins. The enrichment perform a complete analysis and a PPI network
      		analysis. During the enrichment steps, the only protein analyzed are the
      		up- or down-regulated, discriminated by the following parameters.
      	</p>
      	<p>The example figures and tables are from case study on MCF7 cells
      		experiment. Additional details regarding the experiment can be found in
      		section <a href="#example-case-study---proteomics-of-mcf7-cells">3.
      		Example case study - Proteomics of MCF7 cells</a>.
      	</p>
      	
      	<div id="set-settings-for-the-execution-and-read-the-raw-data-from-loaded-files" class="section level3">
      		<h3>1.1. Set settings for the execution and read the raw data from
      			loaded files
      		</h3>
      		<p>ProTN analyse the results of Proteome Discoverer and MaxQuant. The essential parameters and files to run ProTN are: 
      		(additiona details on the input can be found in section <a href="#parameters-required-before-the-execution">2. Details on the input parameters and files</a>)</p>
      		<ul>
      			<li>
      				<p><code>Title of Analysis</code>: title of the experiment. It will
      					be the title of the web page report.
      				</p>
      			</li>
      			<li>
      				<p><code>Software Analyzer</code>: determine with software was use
      					to identify peptides and proteins. <strong>PD</strong> for Protein
      					Discoverer, <strong>MQ</strong> for MaxQuant.
      				</p>
      			</li>
      			<li>
      				<p><code>Input file</code>: file with the information about the
      					samples and the correlation between replicate ID and condition.
      					<button id="preview_input" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Input_Example">[Example of an Input file HERE...]</button>
      				</p>
      			</li>
      			<li>
      				<p><code>Peptides file</code>: raw file of peptides obtained from PD
      					or MQ (file <strong>peptides.txt</strong>).
      				</p>
      			</li>
      			<li>
      				<p><code>Proteins file</code>: raw file of protein groups obtained
      					from PD or MQ (file <strong>proteinGroups.txt</strong>).
      				</p>
      			</li>
      			<li>
      			<p><code>Design for the comparison file</code>: Excel file
      					containing the formulas of the contrast comparison you want to analyse. <button id="preview_design" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Design_Example">[Example of Design file HERE...]</button>
      			</p>
      		</li>
      		</ul>
      	</div>
      	<div id="normalization-of-the-intensities" class="section level3">
      		<h3>1.2. Normalization of the intensities</h3>
      		<p>The intensities are log2 transformed and normalized with DEqMS <span class="citation">(Zhu 2022)</span>. Two methods are applied because a
			double normalization is required, one for peptides and one for proteins.
			At the peptide level, the normalization is done by the function
			equalMedianNormalization, which normalizes intensity distributions in
			samples so that they have median equal to 0.
		</p>
      		<p>At the protein level, this operation is executed by the function
      			medianSweeping. It applies the same median normalization used for
      			peptides, but also it summarizes the peptide intensities into protein
      			relative abundance by the median sweeping method.
      		</p>
      	</div>
      	<div id="imputation-of-the-intensities" class="section level3">
      		<h3>1.3. Imputation of the intensities</h3>
      		<p>After the normalization the intensities from Mass Spectrometer need
			to be imputed. The principal method is based on the PhosR package <span class="citation">(Kim et al. 2021)</span> that performs a complex and
			well-balanced imputation of the data based on the conditions of the
			samples. As a backup method in case of problems with PhosR imputations,
			ProTN uses a Gaussian round imputation.
		</p>
		<p>In this step, two MDSs and two PCAs (proteins and peptides) are
			generated.
		</p>
	</div>
	      <div id="batch-effect-correction" class="section level3">
		<h3>1.4. Batch Effect correction</h3>
		<p>If required by the experiment, a batch correction step can be applyed
			using <a href="https://www.bioconductor.org/packages/release/bioc/html/proBatch.html" style="font-size: 11pt;">proBatch</a> <span class="citation">(Cuklina et
			al. 2018)</span>. The batches need to be defined in the
			<strong>Input</strong> file where column <strong>MS_batch</strong> is
			required.
		</p>
	</div>
      	<div id="statistical-differential-analysis" class="section level3">
      		<h3>1.5. Statistical differential analysis</h3>
      		<p>The workflow continue with the differential analysis. This phase is
			applied to proteins and peptides, to obtain the significant proteins and
			peptides. However the two methodologies are slightly different since for
			proteins is used the DEqMS package <span class="citation">(Zhu
			2022)</span>, instead, for peptides is used the usual Limma package
			<span class="citation">(Ritchie et al. 2015)</span>. DEqMS is developed
			on top of Limma, but the method estimates different prior variances for
			proteins quantified by different numbers of PSMs/peptides per protein,
			therefore achieving better accuracy.
		</p>
		<p>Limma and DEqMS calculate DEPs for each comparison in the
			<em>design</em> file parameter. Each peptide or protein has different
			parameters: the log2 Fold Change, the P.Value, the adjusted P.Value and
			the log2 expression. In this pipeline, a protein/peptide is significant
			if passing 3 thresholds. A protein/peptide for each comparison can be
			Up-regulated or Down-regulated. It is Up-regulated if:
		</p>
		<ul>
			<li>
				<p>the log2 FC is higher than the Fold Change threshold (FC &gt;
					Log2 FC thr),
				</p>
			</li>
			<li>
				<p>the P.Value is lower than the threshold (P.Value &lt; P.Value
					thr),
				</p>
			</li>
			<li>
				<p>the log2 expression is higher than the threshold (log2 expression
					&gt; Signal log2 expr thr).
				</p>
			</li>
		</ul>
		<p>It is Down-regulated if:</p>
		<ul>
			<li>
				<p>the log2 FC is lower of the Fold Change threshold (FC &lt; -Log2
					FC thr),
				</p>
			</li>
			<li>
				<p>the P.Value is lower than the threshold (P.Value &lt; P.Value
					thr),
				</p>
			</li>
			<li>
				<p>the log2 expression is higher than the threshold (log2 expression
					&gt; Signal log2 expr thr).
				</p>
			</li>
		</ul>
		<p>In the output, for each comparison, this distinction is reported in
			the “<em>class</em>” column, which assumes value “+” if is up-regulated,
			“-” if down-regulated and “=” if it is not significant.
		</p>
		<p>Various figures are generated, first a bar plot that graphically
			represents the DEPs identified. Followed by comparison-specific volcano
			plots.
		</p>
      	</div>
      	<div id="enrichment-analysis-of-the-differentially-expressed-proteins" class="section level3">
      		<h3>1.6. Enrichment analysis of the Differentially Expressed
      			Proteins
      		</h3>
      		<div style="display: flex;">
      		<p>The biological interpretation of the Differentially Expressed
			Proteins starts with the enrichment step. To execute this analysis,
			ProTN uses <a href="https://maayanlab.cloud/Enrichr/">EnrichR</a> <span class="citation">(Jawaid 2022)</span>, a widely used tool that search on
			a large number of data sets to obtain information about many categories.
			EnrichR organises its hundreds of data sets in 8 sections:
			Transcription, Pathways, Ontologies, Diseases/Drugs, Cell Types, Misc,
			Legacy, and Crowd.
		</p>
		<p><a href="https://maayanlab.cloud/Enrichr/"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAPAAAACWBAMAAAFl0ShwAAAAMFBMVEXHHwT///8AAAA+Ly9oYmGcmJffrqfFSTvYz870393q6Ojy7Nj38/PUhXv9/Of7+vo64JwrAAAHZUlEQVR42u2af2wTVRzAX//ZQXThXrf+TGuz22oM4Q/LtdTwD45WNhOM1m5uEow0ikUCoU1Z31Tqchk4NBC2mICamFgFIYYl/kcMMdoYl6gjrDFYMRDCP/qvx8T5Y2Z+37u7bv252Slz4b2lu9779nOfez/v3rUIlyb03+/b1LJ9jxTRdmJ032JT2rCEm7BoEnAU4ZZruA1TZAzn1SxqCfpgX/+8qiL8YNnxPXe6PP/1fkvZfmtSou9MWGT7M6/msHmjmDQJYsLchvDgQxibzSI2mUTRjFGr7CvZx5Gy4+fuaHlWW9hTN+zxdS7IctN/YqwY/kJO+X2sr0Idu+jWAeEx2Kq0259JbsJYkrA5YqaUSNtTNImq2Qe0c2TXa1sw7elm+KMR/B49OIRVCNv2mOjxJClSDFuRqRjGLclVXOc8vNrC2bph6/a6YVt4YdYBNogWho/foKNET01l4ZA/FRqjY8ypv2CsiXTrYsPAfeOVLB1Etzo24FMCXovFBBYFAZvbafj2lp4hNgRv4RAexRBuogf/GYYUDf90vNXEhuAtHNXCJsyGKQvbfzp6KFYz3BIKyjARl4SFBfS1vpvaAJ4Piyg3f/BppW6dd9VvEjvvqf+rMIc5fLfDntTlxmG3L6Qu/vnJXFXYluosqHSulaRIGSE2z38+VgU+Eu7L2JKFWSwtcqaVsKMvT44QWfbndNgj3WIn4GgWURKLayFrBiG44qA4QvR+TWqR2nT4SmAmHyWfsUvE7OQkhduwuQNgRDkKu1GO3lfCAdyUkbyG2ZV6IzDd2cUupJIsyxRWsaWdwqoO/7xOO9M8vGjVKDrswSR6Lj2tXV6M08Y6jHV41FQsM4NVDY7KMhl84p2BSF2YlbsCni1MDJ4aCG/Ta1AKh8NKFdiB1py8t9KMC5bgPcXq74UENyS9cD8AL88L7BZkH/2//zqU8wW4Aad5vcWmCk2vW0bfDmX5ZMBhDnOYwxzmMIfvelhZBmwPLAOe+TbSOHx2x1LUX1eH4w99UKvUM3njnQvVgFNKGp7r0wVKOTxlWgw+NdA/EqkO76oPe3rCfUPxt9K5Vm/dIleF4W4/mJL9oc552KiaSSPDkWPw77livgZb4q9n+lKyHMA63Lp+Kzt7cd0U0k97F0ImgD9HCOrV4n0fligaHJ0YPDuR+eYYLsLt7bLkA7gJJYBL0qWRkEgCjAi6l8KSnNVg58CW+LEAoU/1W72TdF3VCiuyJ9sAZiswCtMlYQ5gE3tngXWX8RjTF/przzBhSqjsDrqBx/MbKIx1WKtxF5yzW6Bw1oDd/lZf9FxQg2FRCBs4gHm9sQSlYCJv1LYGF2vbmuocub2ZLcqMMhtwk2FGpXC7Ac9ZM5vcQ9G0ugD2lsCmmrDn6CfEli9YfLgmXNNsD8FysJ+4B7J14YSpGjwcABiWro/juvDHa6vB5zO2ocJAeAfWewesJauVWURXR5KVZT6bfz5eHLD6qKo0Y0SX7BXwH2umg8niTEYXorDphw88C434Is2coA8K7Ne/w2wR6oCX8ymjqb6IyMmGp15n4M184/P2Bw8rjcPO4HKuGMrdepXkMIc5zOFVB3MxF3MxF3MxF3MxF3MxF3MxFxeT57ntnhURj6fknZ4VELsK6dPysysgthO/0oDZpVbJIyS2dPGvJOVXzuzXCEtYT8oiXitClQ5XtcxaYtVBMj1ePELYz5vY89lqv+sqT26kPaVuUOyZ/ej8n4QcGBfGjgsnmLjjKEvZxao6nsaNi4/AQ3R4Ek4O9nYJyrjQQ7/n8S6jBy1N7HFGhf4eEAdJSmbpUVwmvh3eiQ9DwPhl9wXoOyOEJFmXJEalTBBIVw2xFXYyesezhrvxbVneVCp2htV4RnmXxD4lpCck+7dlcbn4F+n+razF9cxRdAi+KWKNa0XCmHb+U4gmQdHEBbaHtL5pkbzTQD9QKh6RAzlq3t1PMjeNzHKxJHXDt09GXxuFQ2bePpnVevWYkZe+/NX+jF5i1DyGLyG0RhfDV0Z7z50oETvI3lBgLp7JHoZGVubFEqt1v6qLfXS7VdtQyUvF4aSV2IHQoYVtzL5RuqBtqNhb0cYTqW5nKHD5jJ+aY+q8mKUOvcTtrCGflDbrYiE3P46Z2GZUuS5+SRtrgqqLIxXiH4kM5v5c1J8dDsYWVPXc5CT7Qo6JO9RScRMuK/F7qLmkV+dLxe1KhdhKDsrdrqm0GvUfb87VamOt5HXEo2XiWJlYrRDbSebDZ763F9KeqPwyriX2/qvi33phMu6FCf1XchXM9sBc42IbalaXKj6sTRcwb/hD8iO7E4EcblxccrGoL/7yMfLDRGKgt4uk2unA2bFwZvvHYjylvTmZW0JVj5MfrGTfJdJ838WL50un1NKr01LEVm2qEpYiVsfJNefGR4hws+JiXku8vlKsz1zYzuZM1tLzYlSjV49f2WmRg4KC/500N7fkWx+417HIQhbfsVQcx6flpxPNK3JffVp+ML8iYnW4m6+duJiLuZiLuZiLuZiLuZiLG0p/AzndS9nlGNYyAAAAAElFTkSuQmCC" width="200" /></a></p></div>
		<p>Since the analysis that a user wants to perform can be different,
			each comparison has 3 sets of proteins: the Up-regulated (called Up),
			the Down-regulated (called Down), and the set (called all) obtained by
			the merge of Up- and Down-regulated proteins. EnrichR provides for each
			term much information. It returns the statistical parameters, like
			P.Value, fdr, odds ratio and combined score. At the same time, it gives
			the overlap size, the number of genes in the term and the genes of the
			input list find in the term.
		</p>
		<p>The tons of data downloaded from EnrichR are exported in two
			modalities. The tool creates an <em>RData</em> of the complete data
			frame, in this way the user can easily import in R to perform further
			analysis and plots. But also, it generates an <em>Excel</em> file with
			only the significant terms. The process uses the filters define in the
			options file to differentiate the relevant terms from the globality.
		</p>
		<p>A term to be significative need to have:</p>
		<ul>
			<li>
				<p>a P.Value lower of P.Value thr for enrichment (P.Value &lt;
					P.Value thr for enrichment),
				</p>
			</li>
			<li>
				<p>an Overlap Size higher than Overlap size thr for enrichment
					(Overlap Size &gt; Overlap size thr for enrichment).
				</p>
			</li>
		</ul>
		<p>Besides that, the pipeline plots the enrichment results in different
			figures. The 4 plots can be split into 2 categories. The first two are
			filtered on the data-sets, they illustrate only the data-sets written in
			the column DB to analyse of the options file. The other 2 are filtered
			on words to search on the term description, it uses the list of words in
			column Terms to search.
		</p>
		<p>Additional details on the output can be found in section <a href="#details-on-the-output-files">4. Details on the output
			files</a>.
		</p>
      	</div>
      	<div id="protein-protein-interaction-network-analysis-of-differentially-expressed-proteins" class="section level3">
      		<h3>1.7. Protein-Protein Interaction network analysis of Differentially
      			Expressed Proteins
      		</h3>
          <div style="display: flex;">
      		  <p>Last analitical step is the Protein-Protein Interaction (PPI) network
      			analysis, since PPIs are essential in almost all processes of the cell,
      			and it is crucial for understanding cell physiology in different states.
      			For each comparison, ProTN analyses the interaction between the DEPs
      			using <a href="https://string-db.org/">STRING</a> <span class="citation">(Szklarczyk et al. 2021)</span>.
      		</p>
      		  <p><a href="https://string-db.org/"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUQAAABaCAMAAAD3n9+rAAABDlBMVEX////////19fXy8vLt7e3s7Ozr6+vm5eXk5OTi4uLh4eHb29vZ2dnX1tfW1tbPz8/MzMzLy8vNysrGxsbCwsLDwcHAwMC8vLy6urq4uLi5t7ezs7OjrNiurq6vrKysrKyfp8+pqamoqKibosakpKSmoqKgoKCXnb2cnJydmpqTmbWampqYmJiXl5eWlpaQlK2UlJSUkZKRkZGMj6WQkJCNjY2Iip6MiYqEhZaFhYWEhISDg4OAgZCEgYKCgoKBgYGAgIB8fYp+fn58fHx7e3t7ent5eXl4eHh0dX51dXV0c3Rzc3NycnJxcXhxcXFwcHBvb29ubXJubW1tbW1ra2tqamxqamppaWloaGhnZ2dyKCOcAAAAAXRSTlMAQObYZgAACl1JREFUeNrtnP932sgRwIVccKFEpqCUrjFpaDicHK1TN1Ji1bHvfIexz70I4gu9zv//j/QJdqXZ1a60K+S+l3c7v/iBEJY+zOx8XTmOFStWrFixYuVrlWa32+1aDPtIN4RE1jPXsqgo7gzg08XFxc0G1kcWRzW5hM/fniTy8gJgYHlUkQl8enlC5R2sm5ZIBZcCm5ThyckFzCySKor47gTJZ7DOpcqK+BJD/GhXxQoCnzDDk29hZJmYQ7y3EPeH+JmD+NZCrCAhvOLXRBtvm8sQLhDDl5u19c4Vkr4vcJpBvIG/VfkSj8yDJPuOginxvqrbb/lkHgQxwCIIxsSrpkHtS4DNaRZrg/ma2BrHwMm8R5dbDSG7j3Y8QoIgWNB3kxvq5P8T2R3llvSdqH65/AlYetNYvJ5o3DE35hWEr2HzfhsrvrqH9XAFl0aZnzuWoIkMIRLZoXjqaUNcmENskVh+SVNDUz4DGDrOCGBzf3HzCeC67TRDWBnE252F+jr2hJhoZEsTItNobYguKTMOTemu4LK9zZ9//u/29L9u3x4BnOmuDT35r9mrCSKArwkRWkYQFddd8E3KrBkmLi1CXLrd9hErPrSXsNQrc3fYtcRz4nmeR6bR9pVTG0QYa0IMTCDiJSiaEt/zPJ+QIM6WIm2PsjpKw5xh8me1Yro+0/MvLrXlmGSa6/oRW1UIL/Rm+Tc9TMen3tH1SCTTxSKIvM4WQnTnOR/IlIJE3K+m4VHcrAix9SVn0GbvHGn5F3pTkWAAnlegFaTgi7h1n91p3NKDGLuaEN30B5pKLNdrGXoU9gout38HMMnKjDr+JVZf/d4QHcfPeUsVxDhn+EUQ2a8TGQczR5MwDGcDwaM4FN6IwlyiMzL/0hyFYRgOc5rpF9hRDRBTim4ZRBqoeloQx9UiGcfp3tEz7444j7KT1IyRPWf+JVkgdzIRVG5aGKLtD5HpjF8Gka63kQ5ET+KwtGQAcP82qXPdAMywRxEUcMC7k61/ad/Bw/Fho3H44gGueYqB2S9qDrEj6IwSohsrvjt3Av2kuR4OYPOW5nWnG/iCPIqwFLLF0UH+5T9wftDYysG5QNEwNjWH6Cz46EUFMaBH4lYpRFIUEBV5kTUqMZxu4C/CcWTE1E1n0vwJzhupfECe5/8Cccq/r4ZIFSwog9iKFbTLZMQVu96J/TzsTobIZ9Mw/OEgg3jwyLWx4KnNWXxfDZEd6pVAJFUyu0Tu+FbUZ3AU1rxNXcRq44sGkm84yNFTOxYDiNTyxXBLPMEsKMOaxndRPgKf0s3wa9GeQzjEEJ9xajwvLkPVBjHSgehLna5wQq+iZ3a6nDXvaoYG0uAEwnwct3CfDOLO/891IFK7EH5S4YRphRIDhfhRgHgZIvkOfkKvIvg3PhjGRRBdFvq7TwVRiFwKIXogCRaFE2IwLTEozPmGN+cZ35la8hMQZ7w593mvxIL/hfc0EH1BcZS5M15diBpip6o1J46F6+dt1tzRtHTD6mMD3rF8gyG+4Y+6cVEqvz9EIUwsgdiCfPjCn+DLXbheiIPt+T2vTEdCxNOGMz7EeeRCHKEX6KFafqt2iHQF62hClAXSUu/eqgDRXaOZpdMNXzCYiX1mXjPdBXzIIF7lqoy+pD9VF8RpzvhKILLyZk8FMSjuWxVXcADeswGHTVJIUFuzQLW7hF/gA0v7ruC6uM6+IG5tEL0or1clEFkEg8JAGcRFJYjbAsS7Vyev3t4DDLq43ipYr2DfI4DZ767h8U2/0ei/eRQLELuVKOI6dKQOiKiyHbgGEJmqjRWH42p5M2MVslJYUttC9dZJfpIuaxKE23KPO6LnrhUdA771qPbU5RCLOywaEFtiG5o/rCo+iP9VdQvt0VkYXlJTHQLQjQLL/Ewny2Ayjb2D12E4GShN1eUxjmuDGHQcM4hpw+JpIPKm2l7uStt5a04+NGG9QJXJ58UP8L27tUDMl+/LIdKoKP0vTwARmerZ1qQl1ryr6jRR5XaiNTyLh0miejQx8swhenyw+BQQkevdmvRSNqF9Bu0j3Atcao5xI6se17Qmjo0hsrxlLjscKUIcM4jY9baX8IPUUAfwL9x21rFmsc0pvQxt7+z2Uq0eG0NkWZRnECd6qWjVpXBYmPRNX0s+8weAVVeZCBYLKZhHMIkT01GZnilE3rdE3OFxKSMtiHyCciebuhmsAH6vLkmUeRj1QItZsO1LWvJaEJnR+pnu6de1tSByqXITfriEZdsRO3x/x/VrA2tW1lIqpX2SO9aD2EH4eYge8PXJihCd1RKXaIYJM1zxT3rNba5JMDHc2KK+UNPceZFL0vQgMqslOYj00/G+EHF/ftsLGKyQSQ939o2bBEvTLVaL4ka68RhJzxiim3VSBIjzslqYHkTUlqIKlzTzuzh45Jp+ptac6sH+EF3TKo7YSpnmIJZ2nfUgogYpQ+VOdhFNd8kmdJoZuYnxNjVSF0SqNwtziCyWadFymqPMratBRPYcpka7ja1HaUKNbdjYmmuEmKuhakNsMVUkwuFpiSpqQkztGc+MNC/hVzyhk+qfuTVTc65hTWRu1jeHmPIn8qxQGeVoQkztGTftne4XrlabsjO3Zhqm1eCd2RnjChCpe5uKEJmhQ2cviKk9Yz89AfjHCkLsk6tac6uuODG946gKRKpyvniYrYpxZy+IVAORh9mVbHDhxpnsMpsK1jyvKWNBb7sVINLriFRZoYKiLkRKL7PmtGQzykyaZjbF1izbxOXXkztz6tSrAtFFhWJZVii/Eu3BmJ0dM2tmAQ5fzKaVimJrnse5xhSpo4ojnjKtAhGX1xyhgamqVxpA/DP8uIbv4WeHD7VRW8VxnH/Cd4kxhE5J+27aQxx7wX71REe+KC4qQcxUTjjcQdVjwi07rq/rnScA8HBFm1Zc0peZdDNpbN3eAohTtZK8AJLdmduNQNlP7NYEUYgUzSB2FBDRHozdXqCet90MNA70irKO417D44ukkdz/APAnofzATPqPa7h6nnSajx9g3VZ91VTVoouq9lhUHEgliNm+qRyEcfEkXBnEEG7ZXMjxI4iFMGrSv2YTOOeg3EW+UFzC3HXqgihUMwwhpr5F4hMXe0AcZAwbjWP4RXq/P+IppnPl44ZaUlVc9IpTKjOIY86eDSGmC460eqzEuBiXxMZ33KjcuTSCacMD+szBAyj3p7VIJGqhX1arM4PIh+6mEIuHb7xxnmM8J6XjTm08ntRoHEq97wiO8Yde5JZNzmQ8Mt1tmA8C0vvanh/R6hG63T9ONvv3tObFhjyfxsNaumwKc9pnjhVOyfocxCvjOW0rEohhXmILscQ5c9tSGo930rTwGT+nPbHgsDThFvN5Lg1fhDntc/scu5zTeI743ILsUQ/uGs9pH4J9ZpOY1MHjM6RjoWLlzCLyg9vCCOe36loe++nOW5WOXcMtjXKe3doARyITgKvjfj8Zwb5T1RbcEOD8eb///NzgETm/LYu+VjyNgHMuazrjbZ2KKvkbjUajQYmGHSUfsk/St2LFihUrVqxYsWJFLv8D5jNcr4auNZwAAAAASUVORK5CYII=" width="270"></a></p>
      		</div>
      		<p>The species-specific database is retrieved from STRING server, an
      			accurate analysis discover all the interactions and an iGraph <span class="citation">(Csardi and Nepusz 2006)</span> network is generated.
      			Later, the proteins are clustered via iGraph function which identify
      			dense subgraph by optimizing modularity score.
      		</p>
      		<p>Since the network can vary a lot on composition, two ggplot layout
      			are used: Fruchterman-Reingold algorithm and the Kamada-Kawai
      			algorithm.
      		</p>
      	</div>
      	<div id="report-creation-and-download-of-the-results" class="section level3">
      		<h3>1.8. Report creation and download of the results</h3>
      		<p>The results are summarized in a web-page HTML report. Other than
      			this, the experiment is described by a large number of files, a
      			description of each file generated can be found in section <a href="#details-on-the-output-files">4. Details on the output files</a>.
      			All the files are group in a <em>zip</em> file and downloaded.
      		</p>
      	</div>
      </div>
    </li>
  </ul>
</div>
<div id="workflow-of-phosprotn" class="section level2">
  <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
        	<h2>2. Workflow of PhosProTN</h2>
  	      <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	      <ul>
      		<li>Workflow phosphoproteomics</li>
      	</ul>
      </div>
    </li>
  </ul>
</div>
<div id="parameters-required-before-the-execution" class="section level2">
	 <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
      	  <h2>3. Details on the input parameters and files</h2>
  	      <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	    	<!--TODO: ADD DIA analysis-->
		  <ul>
			<li>
				<p><code>Title of Analysis</code>: title of the experiment. It will
					be the title of the web page report.
				</p>
			</li>
			<li>
				<p><code>Brief Description</code>: description of the current
					experiment. It is the first paragraph of the report.
				</p>
			</li>
			<li>
				<p><code>Software Analyzer</code>: determine with software was use
					to identify peptides and proteins. <strong>PD</strong> for Protein
					Discoverer, <strong>MQ</strong> for MaxQuant.
				</p>
			</li>
			<li>
				<p><code>Input file</code>: file with the information about the
					samples and the correlation between replicate ID and condition. The
					<strong>Input</strong> file is an Excel file with the following column
					<button id="preview_input" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Input_Example">[Example of an Input file HERE...]</button>:
				</p>
				<ul>
					<li><strong>Condition</strong> column (<code>REQUIRED</code>): define
						the condition of each sample that divide the samples in groups. The
						conditions need to be the same of the the Contrast Design. In the
						example the two condition are DMSO and TRT.
					</li>
					<li><strong>Color</strong> columns (<code>OPTIONAL</code>): define a
						color for the samples in the graphs. If not present use a default
						palette.
					</li>
					<li><strong>MS_batch</strong> columns: define the groups of batch in the
						samples. <code>REQUIRED FOR BATCH EFFECT CORRECTION</code>.
					</li>
					<li>
						<strong>Sample</strong> column: define the names for the samples.
						<ul>
							<li>In case of <strong>PD</strong> files use the <em>Input</em> file
								obtained from PD, an example of PD Input file is present in
								<em>file_input_raw/Input_filter.xlsx</em>, the <strong>Sample</strong>
								column in optional, if is not present the software extract the names for
								the <strong>“File Name”</strong> column.
							</li>
							<li>In case of <strong>MQ</strong> analysis the this column is
								<code>REQUIRED</code>
							</li>
						</ul>
					</li>
				</ul>
			</li>
			<li>
				<p><code>Peptides file</code>: raw file of peptides obtained from PD
					or MQ (file <strong>peptides.txt</strong>).
				</p>
			</li>
			<li>
				<p><code>Proteins file</code>: raw file of protein groups obtained
					from PD or MQ (file <strong>proteinGroups.txt</strong>).
				</p>
			</li>
			<li>
			<p><code>Design for the comparison file</code>: Excel file
					containing the formulas of the contrast comparison you want to analyse.
					The table can have 3 columns <button id="preview_design" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Design_Example">[Example of Design file HERE...]</button>:
			</p>
			<ul>
				<li><strong>Formule</strong> column (<code>REQUIRED</code>): The
					formulas need to follow the syntax of <a href="https://bioinf.wehi.edu.au/limma/">Limma</a>.
					<code>AT LEAST 1 FORMULA IS REQUIRED</code>.
				</li>
				<li><strong>Name</strong> column (<code>OPTIONAL</code>): personalized
					name assign to the comparison.
				</li>
				<li><strong>Color</strong> columns (<code>OPTIONAL</code>): define a
					color for the condition in the graphs. If not present use a default
					palette.
				</li>
			</ul>
		</li>
		<li>
			<p><code>OPTIONAL</code>:</p>
			<ul>
				<li>
					<p><strong>Signal log2 expr thr</strong>: signal log2 expression
					threshold for the differential analysis. <em>DEFAULT:</em> DEPs if
					<code>Signal log2 expression threshold &gt; -∞</code> (No Limit,
					represent by value “inf” in the cell)
				</p>
			</li>
			<li>
					<p><strong>Log2 FC thr</strong>: Fold Change threshold for the
					differential analysis. <em>DEFAULT:</em> DEPs if
					<code>log2 Fold Change threshold = 0.75</code> (Up-regulated &gt; 0.75,
					Down-regulated &lt; -0.75)
				</p>
			</li>
			<li>
					<p><strong>P.Value thr</strong>: p.value threshold for the
						differential analysis. <em>DEFAULT:</em> DEPs if
					<code>P value threshold &lt; 0.05</code>
				</p>
			</li>
			<li>
				<p><strong>Batch Correction</strong>: execution of the batch effect
					correction performed by <a href="https://www.bioconductor.org/packages/release/bioc/html/proBatch.html">proBatch</a>.
					<span class="citation">(Cuklina et al. 2018)</span>.
					(<strong>TRUE</strong> or <strong>FALSE</strong>). If
					<strong>TRUE</strong>, column <strong>MS_batch</strong> required in
					<strong>Input</strong> file.
				</p>
			</li>
			<li>
				<p><strong>Control Boxplot proteins</strong>: list of proteins used as
					control of the intensities. For each protein a boxplot is generated
					comparing the mean of the intensities group by condition.
				</p>
			</li>
			<li>
					<p><strong>Execute Enrichment</strong>: boolean value for the
						execution of the enrichment step. (<strong>TRUE</strong> or
					<strong>FALSE</strong>)
				</p>
			</li>
			<li>
				<p><strong>Execute PPI network STRINGdb</strong>: boolean value for the
					execution of the network analysis. (<strong>TRUE</strong> or
					<strong>FALSE</strong>)
				</p>
			</li>
			<li>
				<p><strong>P.Value thr for enrichment</strong>: p.value threshold of the
					enriched terms. <strong>DEFAULT:</strong> Term is significant if
					<code>P Value threshold for Enrichment &lt; 0.05</code>
				</p>
			</li>
			<li>
				<p><strong>Overlap size thr for enrichment</strong>: Overlap size
					threshold. The overlap size is the number of DEPs discovered in the
					enriched terms. <strong>DEFAULT:</strong> Term is significant if
					<code>Minimum number of overlap genes with enriched terms &gt; 5</code>
				</p>
			</li>
			<li>
				<p><strong>Enrichment</strong>: determine of filter the plot of the
					enrichment results. The enrichment can be filter by terms or by DBs. In
					the column <em>“Terms to search”</em> you can write the word that you
					want to search in the results of EnrichR (EX: MYC, C-MYC, Senescence,…).
					Instead, in the column <em>“DB to analyse”</em> you can write the DBs
					that you want to see in your plots (EX: ChEA_2016, KEGG_2021_Human,
					BioPlanet_2019, GO_Biological_Process_2021,…), using the same name that
					you can find in EnrichR. If the two columns are empty, no plots of
					enrichment are returned.
				</p>
			</li>
		</ul>
		</div>
    </li>
  </ul>
</div>
<div id="example-case-study---proteomics-of-mcf7-cells" class="section level2">
  <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
        	<h2>4. Example case study - Proteomics of MCF7 cells</h2>
        	<hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	      <p><code>Proteomics of MCF7 cells (LC-MS/MS Orbitrap Fusion Tribrid)</code></p> 
	      <p>PRIDE:	<a href="http://proteomecentral.proteomexchange.org/cgi/GetDataset?ID=PXD009417">PXD009417</a>.
      	<p>Clamer M, Tebaldi T, Lauria F, et al. Active Ribosome Profiling with RiboLace. Cell Rep. 2018;25(4):1097-1108.e5.</p>
      	<p>Its execution is performed clicking the button <strong>Case Study Example</strong>.</p>
      </div>
    </li>
  </ul>
</div>
<div id="details-on-the-output-files" class="section level2">
  <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
        	<h2>5. Details on the output files</h2>
  	      <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	      <ul>
      		<li><code>report.html</code>: complete report of the analysis with all pics and results of the enrichment.</li>
      		<li><code>database_env_R.RData</code>: RData containing the	essential variables for further analysis.</li>
      		<li><code>normalised_intensity_table.xlsx</code>: excel file with
      			the matrix used in the analysis. In sheet <em>Normalized proteins matrix
      			intensity</em> is written the normalized and imputed version of the
      			original intensities of each protein in each sample. In sheet
      			<em>Normalized peptides matrix intensity</em> is written the normalized
      			and imputed version of the original intensities of each peptide in each
      			sample. In sheet <em>Normalized proteins intensity group by</em> is
      			written the average and the standard deviation of each condition
      			described in the Input file for each protein. In sheet <em>Normalized
      			peptides intensity group by</em> is written the average and the standard
      			deviation of each condition described in the Input file for each
      			peptide.
      		</li>
      		<li>
      			<code>DEPs_table.xlsx</code>: summary of the result of Limma. In
      			sheet <em>DEPs_proteins</em>, Column <em>id</em> is the name of the
      			protein. In sheet <em>DEPs_peptides</em>, column <em>Accession</em> is
      			the UniprotID of protein, column <em>Description</em> is the description
      			of protein, <em>GeneName</em> is Gene Symbol, <em>Annotated
      			Sequence</em> is the sequence of peptides, <em>Modifications</em> is the
      			modification that affect the peptides, <em>Position in Master
      			Proteins</em> is the UniprotID with the position on the sequence. For
      			each contrast rule are present 5 columns:
      			<ul>
      				<li><em>class</em>: is “+” if the protein is up-regulated following the	parameters set, “-” if the protein is down-regulated, “=” otherwise</li>
      				<li><em>log2_FC</em>: is the log2 Fold Change of the contrast</li>
      				<li><em>p_val</em>: is the p. value</li>
      				<li><em>p_adj</em>: is the adjusted p. value</li>
      				<li><em>log2_expr</em>: is the signal log2 expression</li>
      			</ul>
      		</li>
      		<li><code>MDS</code>
      		  <ul>
      		    <li><code>general_mds_proteins.pdf</code>: MDS of the samples by proteins.</li>
      		    <li><code>general_mds_peptides.pdf</code>: MDS of the samples by peptides.</li>
      		    <li><code>corrected_mds_proteins.pdf</code>: MDS of the samples by proteins after the bacth effect correction.</li>
      		    <li><code>corrected_mds_peptides.pdf</code>: MDS of the samples by peptides after the bacth effect correction.</li>
      		    <li><code>DEPs_proteins_mds.pdf</code>: MDS of the DEPs discovered based on proteins.</li>
      		    <li><code>DEPs_peptides_mds.pdf</code>: MDS of the DEPs discovered based on peptides.</li>
      		  </ul>
      		</li>
      		<li><code>PCA</code>
      		  <ul>
      		    <li><code>general_pca_proteins.pdf</code>: PCA of the samples by proteins.</li>
      		    <li><code>general_pca_peptides.pdf</code>: PCA of the samples by peptides.</li>
      		    <li><code>corrected_pca_proteins.pdf</code>: PCA of the samples by proteins after the bacth effect correction.</li>
      		    <li><code>corrected_pca_peptides.pdf</code>: PCA of the samples by peptides after the bacth effect correction.</li>
      		    <li><code>DEPs_proteins_pca.pdf</code>: PCA of the DEPs discovered based on proteins.</li>
      		    <li><code>DEPs_peptides_pca.pdf</code>: PCA of the DEPs discovered based on peptides.</li>
      		  </ul>
      		</li>
      		<li><code>boxplot_control_intensity.pdf</code>: boxplot for the control of the intensities of selected proteins.</li>
      		<li><code>DEPs_proteins_count_barplot.pdf</code>: number of DEPs find for each contrast design based on proteins. Light grey are the
      			down-regulated, dark grey are the up-regulated genes.</li>
      		<li><code>DEPs_peptides_count_barplot.pdf</code>: number of DEPs find for each contrast design based on peptides. Light grey are the
      			down-regulated, dark grey are the up-regulated genes.</li>
      		<li><code>DEPs_enrichment_data.RData</code>: result of the enrichment made with EnrichR.</li>
      		<li><code>DEPs_enrichment_table.xlsx</code>: excel with only the enriched terms with an high significance (P.Value &lt; 0.05, Overlap Size &gt;= 4).</li>
      		<li>
      			<code>enrich</code> directory: plots of the enrichment data	under the parameters selected
      			<ul>
      				<li><code>DBs_enr.pdf</code>: dot matrix of the enrichment result	divided in up- and down-regulated genes. Filtered by the datasets choose.</li>
      				<li><code>terms_enr.pdf</code>: dot matrix of the enrichment result divided in up- and down-regulated genes. Filtered by the terms written.</li>
      				<li><code>DBs_enr_all.pdf</code>: dot matrix of the enrichment result combining the up- and down- regulated genes in a single list.
      					Filtered by the datasets choose</li>
      				<li><code>terms_enr_all.pdf</code>: dot matrix of the enrichment result combining the up- and down- regulated genes in a single list.
      					Filtered by the terms written.</li>
      			</ul>
      		</li>
      		<li>
      			<code>network</code> directory: plots of the up- and down-
      			regulated genes based on STRINGdb. For each contrast design are create 2
      			files:
      			<!-- - **Prefilter**: histogram representing the number of genes in each community discovered. -->
      			<ul>
      				<li><code>Postfilter</code>: histogram representing the number of
      					genes in each community discovered after a filter on the genes and on
      					the links.
      				</li>
      				<li><code>Graph</code>: visual representation of the network
      					discovered in two layouts: <em>fr</em> and <em>kk</em>.
      				</li>
      			</ul>
      		</li>
      	</ul>
      </div>
    </li>
  </ul>
</div>
<hr id="linesubtitle">
<div id="refs" class="references csl-bib-body hanging-indent">
		<div id="ref-igraph" class="csl-entry">
			Csardi, Gabor, and Tamas Nepusz. 2006. <span>“The Igraph Software
			Package for Complex Network Research”</span> Complex Systems: 1695. <a href="https://igraph.org">https://igraph.org</a>.
		</div>
		<br>
		<div id="ref-proBatch" class="csl-entry">
			Cuklina, Jelena, Chloe H. Lee, Evan G. Willams, Ben Collins, Tatjana
			Sajic, Patrick Pedrioli, Maria Rodriguez-Martinez, and Ruedi Aebersold.
			2018. <span>“Computational Challenges in Biomarker Discovery from
			High-Throughput Proteomic Data.”</span> <a href="https://doi.org/10.3929/ethz-b-000307772">https://doi.org/10.3929/ethz-b-000307772</a>.
		</div>
		<br>
		<div id="ref-enrichR" class="csl-entry">
			Jawaid, Wajid. 2022. <span>“enrichR: Provides an r Interface to
			’Enrichr’.”</span> <a href="https://CRAN.R-project.org/package=enrichR">https://CRAN.R-project.org/package=enrichR</a>.
		</div>
		<br>
		<div id="ref-PhosR" class="csl-entry">
			Kim, Hani Jieun, Taiyun Kim, Nolan J Hoffman, Di Xiao, David E James,
			Sean J Humphrey, and Pengyi Yang. 2021. <span>“PhosR Enables Processing
			and Functional Analysis of Phosphoproteomic Data”</span> 34. <a href="https://doi.org/10.1016/j.celrep.2021.108771">https://doi.org/10.1016/j.celrep.2021.108771</a>.
		</div>
		<br>
		<div id="ref-limma" class="csl-entry">
			Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law,
			Wei Shi, and Gordon K Smyth. 2015.
			<span>“<span></span>Limma<span></span> Powers Differential Expression
			Analyses for <span></span>RNA<span></span>-Sequencing and Microarray
			Studies”</span> 43: e47. <a href="https://doi.org/10.1093/nar/gkv007">https://doi.org/10.1093/nar/gkv007</a>.
		</div>
		<br>
		<div id="ref-STRINGdb" class="csl-entry">
			Szklarczyk, Damian, Annika L Gable, Katerina C Nastou, David Lyon,
			Rebecca Kirsch, Sampo Pyysalo, Nadezhda T Doncheva, et al. 2021.
			<span>“The STRING Database in 2021: Customizable Protein-Protein
			Networks, and Functional Characterization of User-Uploaded
			Gene/Measurement Sets.”</span> 49.
		</div>
		<br>
		<div id="ref-DEqMS" class="csl-entry">
			Zhu, Yafeng. 2022. <span>“DEqMS: A Tool to Perform Statistical Analysis
			of Differential Protein Expression for Quantitative Proteomics
			Data.”</span>
		</div>
</div>
