<p>ProTN is an integrative pipeline that analyze DDA proteomics data
	obtained from MS. It perform a complete analysis of the raw files from
	<a href="https://www.thermofisher.com/it/en/home/industrial/mass-spectrometry/liquid-chromatography-mass-spectrometry-lc-ms/lc-ms-software/multi-omics-data-analysis/proteome-discoverer-software.html">Proteome
	Discoverer</a> (PD) or <a href="https://maxquant.org/">MaxQuant</a>
	(MQ), with their biological interpretation with enrichement and network
	analysis. ProTN executes a dual level analysis, at protein and peptide
	level.
</p>
<div id="workflow-of-protn" class="section level2">
  <ul id="collapse_ul" class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
    	    <h2>1. Workflow of ProTN</h2>
    	    <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
      	<p>Later in this section we describes in details the workflow of the
      		pipeline. In the first phases it imputes and normalizes the data
      		obtaining the matrix that is used in the enrichment analysis of the
      		proteins. The enrichment perform a complete analysis and a PPI network
      		analysis. During the enrichment steps, the only protein analyzed are the
      		up- or down-regulated, discriminated by the following parameters.
      	</p>
      	<div id="set-settings-for-the-execution-and-read-the-raw-data-from-loaded-files" class="section level3">
      		<h3>1.1. Set settings for the execution and read the raw data from
      			loaded files
      		</h3>
      		<p>Lorem Ipsum</p>
      	</div>
      	<div id="normalization-of-the-intensities" class="section level3">
      		<h3>1.2. Normalization of the intensities</h3>
      		<p>Lorem Ipsum</p>
      	</div>
      	<div id="imputation-of-the-intensities" class="section level3">
      		<h3>1.3. Imputation of the intensities</h3>
      		<p>Lorem Ipsum</p>
      	</div>
      	<div id="statistical-differential-analysis" class="section level3">
      		<h3>1.4. Statistical differential analysis</h3>
      		<p>Lorem Ipsum</p>
      	</div>
      	<div id="enrichment-analysis-of-the-differentially-expressed-proteins" class="section level3">
      		<h3>1.5. Enrichment analysis of the Differentially Expressed
      			Proteins
      		</h3>
      		<p>Lorem Ipsum</p>
      	</div>
      	<div id="protein-protein-interaction-network-analysis-of-differentially-expressed-proteins" class="section level3">
      		<h3>1.6. Protein-Protein Interaction network analysis of Differentially
      			Expressed Proteins
      		</h3>
          p<div style="display: flex;">
      		  <p>Last analitical step is the Protein-Protein Interaction (PPI) network
      			analysis, since PPIs are essential in almost all processes of the cell,
      			and it is crucial for understanding cell physiology in different states.
      			For each comparison, ProTN analyses the interaction between the DEPs
      			using <a href="https://string-db.org/">STRING</a> <span class="citation">(Szklarczyk et al. 2021)</span>.
      		</p>
      		  <p><a href="https://string-db.org/"><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAUQAAABaCAMAAAD3n9+rAAABDlBMVEX////////19fXy8vLt7e3s7Ozr6+vm5eXk5OTi4uLh4eHb29vZ2dnX1tfW1tbPz8/MzMzLy8vNysrGxsbCwsLDwcHAwMC8vLy6urq4uLi5t7ezs7OjrNiurq6vrKysrKyfp8+pqamoqKibosakpKSmoqKgoKCXnb2cnJydmpqTmbWampqYmJiXl5eWlpaQlK2UlJSUkZKRkZGMj6WQkJCNjY2Iip6MiYqEhZaFhYWEhISDg4OAgZCEgYKCgoKBgYGAgIB8fYp+fn58fHx7e3t7ent5eXl4eHh0dX51dXV0c3Rzc3NycnJxcXhxcXFwcHBvb29ubXJubW1tbW1ra2tqamxqamppaWloaGhnZ2dyKCOcAAAAAXRSTlMAQObYZgAACl1JREFUeNrtnP932sgRwIVccKFEpqCUrjFpaDicHK1TN1Ji1bHvfIexz70I4gu9zv//j/QJdqXZ1a60K+S+l3c7v/iBEJY+zOx8XTmOFStWrFixYuVrlWa32+1aDPtIN4RE1jPXsqgo7gzg08XFxc0G1kcWRzW5hM/fniTy8gJgYHlUkQl8enlC5R2sm5ZIBZcCm5ThyckFzCySKor47gTJZ7DOpcqK+BJD/GhXxQoCnzDDk29hZJmYQ7y3EPeH+JmD+NZCrCAhvOLXRBtvm8sQLhDDl5u19c4Vkr4vcJpBvIG/VfkSj8yDJPuOginxvqrbb/lkHgQxwCIIxsSrpkHtS4DNaRZrg/ma2BrHwMm8R5dbDSG7j3Y8QoIgWNB3kxvq5P8T2R3llvSdqH65/AlYetNYvJ5o3DE35hWEr2HzfhsrvrqH9XAFl0aZnzuWoIkMIRLZoXjqaUNcmENskVh+SVNDUz4DGDrOCGBzf3HzCeC67TRDWBnE252F+jr2hJhoZEsTItNobYguKTMOTemu4LK9zZ9//u/29L9u3x4BnOmuDT35r9mrCSKArwkRWkYQFddd8E3KrBkmLi1CXLrd9hErPrSXsNQrc3fYtcRz4nmeR6bR9pVTG0QYa0IMTCDiJSiaEt/zPJ+QIM6WIm2PsjpKw5xh8me1Yro+0/MvLrXlmGSa6/oRW1UIL/Rm+Tc9TMen3tH1SCTTxSKIvM4WQnTnOR/IlIJE3K+m4VHcrAix9SVn0GbvHGn5F3pTkWAAnlegFaTgi7h1n91p3NKDGLuaEN30B5pKLNdrGXoU9gout38HMMnKjDr+JVZf/d4QHcfPeUsVxDhn+EUQ2a8TGQczR5MwDGcDwaM4FN6IwlyiMzL/0hyFYRgOc5rpF9hRDRBTim4ZRBqoeloQx9UiGcfp3tEz7444j7KT1IyRPWf+JVkgdzIRVG5aGKLtD5HpjF8Gka63kQ5ET+KwtGQAcP82qXPdAMywRxEUcMC7k61/ad/Bw/Fho3H44gGueYqB2S9qDrEj6IwSohsrvjt3Av2kuR4OYPOW5nWnG/iCPIqwFLLF0UH+5T9wftDYysG5QNEwNjWH6Cz46EUFMaBH4lYpRFIUEBV5kTUqMZxu4C/CcWTE1E1n0vwJzhupfECe5/8Cccq/r4ZIFSwog9iKFbTLZMQVu96J/TzsTobIZ9Mw/OEgg3jwyLWx4KnNWXxfDZEd6pVAJFUyu0Tu+FbUZ3AU1rxNXcRq44sGkm84yNFTOxYDiNTyxXBLPMEsKMOaxndRPgKf0s3wa9GeQzjEEJ9xajwvLkPVBjHSgehLna5wQq+iZ3a6nDXvaoYG0uAEwnwct3CfDOLO/891IFK7EH5S4YRphRIDhfhRgHgZIvkOfkKvIvg3PhjGRRBdFvq7TwVRiFwKIXogCRaFE2IwLTEozPmGN+cZ35la8hMQZ7w593mvxIL/hfc0EH1BcZS5M15diBpip6o1J46F6+dt1tzRtHTD6mMD3rF8gyG+4Y+6cVEqvz9EIUwsgdiCfPjCn+DLXbheiIPt+T2vTEdCxNOGMz7EeeRCHKEX6KFafqt2iHQF62hClAXSUu/eqgDRXaOZpdMNXzCYiX1mXjPdBXzIIF7lqoy+pD9VF8RpzvhKILLyZk8FMSjuWxVXcADeswGHTVJIUFuzQLW7hF/gA0v7ruC6uM6+IG5tEL0or1clEFkEg8JAGcRFJYjbAsS7Vyev3t4DDLq43ipYr2DfI4DZ767h8U2/0ei/eRQLELuVKOI6dKQOiKiyHbgGEJmqjRWH42p5M2MVslJYUttC9dZJfpIuaxKE23KPO6LnrhUdA771qPbU5RCLOywaEFtiG5o/rCo+iP9VdQvt0VkYXlJTHQLQjQLL/Ewny2Ayjb2D12E4GShN1eUxjmuDGHQcM4hpw+JpIPKm2l7uStt5a04+NGG9QJXJ58UP8L27tUDMl+/LIdKoKP0vTwARmerZ1qQl1ryr6jRR5XaiNTyLh0miejQx8swhenyw+BQQkevdmvRSNqF9Bu0j3Atcao5xI6se17Qmjo0hsrxlLjscKUIcM4jY9baX8IPUUAfwL9x21rFmsc0pvQxt7+z2Uq0eG0NkWZRnECd6qWjVpXBYmPRNX0s+8weAVVeZCBYLKZhHMIkT01GZnilE3rdE3OFxKSMtiHyCciebuhmsAH6vLkmUeRj1QItZsO1LWvJaEJnR+pnu6de1tSByqXITfriEZdsRO3x/x/VrA2tW1lIqpX2SO9aD2EH4eYge8PXJihCd1RKXaIYJM1zxT3rNba5JMDHc2KK+UNPceZFL0vQgMqslOYj00/G+EHF/ftsLGKyQSQ939o2bBEvTLVaL4ka68RhJzxiim3VSBIjzslqYHkTUlqIKlzTzuzh45Jp+ptac6sH+EF3TKo7YSpnmIJZ2nfUgogYpQ+VOdhFNd8kmdJoZuYnxNjVSF0SqNwtziCyWadFymqPMratBRPYcpka7ja1HaUKNbdjYmmuEmKuhakNsMVUkwuFpiSpqQkztGc+MNC/hVzyhk+qfuTVTc65hTWRu1jeHmPIn8qxQGeVoQkztGTftne4XrlabsjO3Zhqm1eCd2RnjChCpe5uKEJmhQ2cviKk9Yz89AfjHCkLsk6tac6uuODG946gKRKpyvniYrYpxZy+IVAORh9mVbHDhxpnsMpsK1jyvKWNBb7sVINLriFRZoYKiLkRKL7PmtGQzykyaZjbF1izbxOXXkztz6tSrAtFFhWJZVii/Eu3BmJ0dM2tmAQ5fzKaVimJrnse5xhSpo4ojnjKtAhGX1xyhgamqVxpA/DP8uIbv4WeHD7VRW8VxnH/Cd4kxhE5J+27aQxx7wX71REe+KC4qQcxUTjjcQdVjwi07rq/rnScA8HBFm1Zc0peZdDNpbN3eAohTtZK8AJLdmduNQNlP7NYEUYgUzSB2FBDRHozdXqCet90MNA70irKO417D44ukkdz/APAnofzATPqPa7h6nnSajx9g3VZ91VTVoouq9lhUHEgliNm+qRyEcfEkXBnEEG7ZXMjxI4iFMGrSv2YTOOeg3EW+UFzC3HXqgihUMwwhpr5F4hMXe0AcZAwbjWP4RXq/P+IppnPl44ZaUlVc9IpTKjOIY86eDSGmC460eqzEuBiXxMZ33KjcuTSCacMD+szBAyj3p7VIJGqhX1arM4PIh+6mEIuHb7xxnmM8J6XjTm08ntRoHEq97wiO8Yde5JZNzmQ8Mt1tmA8C0vvanh/R6hG63T9ONvv3tObFhjyfxsNaumwKc9pnjhVOyfocxCvjOW0rEohhXmILscQ5c9tSGo930rTwGT+nPbHgsDThFvN5Lg1fhDntc/scu5zTeI743ILsUQ/uGs9pH4J9ZpOY1MHjM6RjoWLlzCLyg9vCCOe36loe++nOW5WOXcMtjXKe3doARyITgKvjfj8Zwb5T1RbcEOD8eb///NzgETm/LYu+VjyNgHMuazrjbZ2KKvkbjUajQYmGHSUfsk/St2LFihUrVqxYsWJFLv8D5jNcr4auNZwAAAAASUVORK5CYII=" width="270"></a></p>
      		</div>
      		<p>The species-specific database is retrieved from STRING server, an
      			accurate analysis discover all the interactions and an iGraph <span class="citation">(Csardi and Nepusz 2006)</span> network is generated.
      			Later, the proteins are clustered via iGraph function which identify
      			dense subgraph by optimizing modularity score.
      		</p>
      		<p>Since the network can vary a lot on composition, two ggplot layout
      			are used: Fruchterman-Reingold algorithm and the Kamada-Kawai
      			algorithm.
      		</p>
      	</div>
      	<div id="report-creation-and-download-of-the-results" class="section level3">
      		<h3>1.7. Report creation and download of the results</h3>
      		<p>The results are summarized in a web-page HTML report. Other than
      			this, the experiment is described by a large number of files, a
      			description of each file generated can be found in section <a href="#details-on-the-output-files">4. Details on the output files</a>.
      			All the files are group in a <em>zip</em> file and downloaded.
      		</p>
      	</div>
      </div>
    </li>
  </ul>
</div>
<div id="parameters-required-before-the-execution" class="section level2">
	 <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
      	  <h2>2. Parameters required before the execution</h2>
  	      <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	    	<!--TODO: ADD DIA analysis-->
		    <ul>
			<li>
				<p><code>Title of Analysis</code>: title of the experiment. It will
					be the title of the web page report.
				</p>
			</li>
			<li>
				<p><code>Brief Description</code>: description of the current
					experiment. It is the first paragraph of the report.
				</p>
			</li>
			<li>
				<p><code>Software Analyzer</code>: determine with software was use
					to identify peptides and proteins. <strong>PD</strong> for Protein
					Discoverer, <strong>MQ</strong> for MaxQuant.
				</p>
			</li>
			<li>
				<p><code>Input file</code>: file with the information about the
					samples and the correlation between replicate ID and condition. The
					<strong>Input</strong> file is an Excel file with the following column
					<button id="preview_input" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Input_Example">[Example of an Input file HERE...]</button>:
				</p>
				<ul>
					<li><strong>Condition</strong> column (<code>REQUIRED</code>): define
						the condition of each sample that divide the samples in groups. The
						conditions need to be the same of the the Contrast Design. In the
						example the two condition are DMSO and TRT.
					</li>
					<li><strong>Color</strong> columns (<code>OPTIONAL</code>): define a
						color for the samples in the graphs. If not present use a default
						palette.
					</li>
					<li><strong>MS_batch</strong> columns: define the groups of batch in the
						samples. <code>REQUIRED FOR BATCH EFFECT CORRECTION</code>.
					</li>
					<li>
						<strong>Sample</strong> column: define the names for the samples.
						<ul>
							<li>In case of <strong>PD</strong> files use the <em>Input</em> file
								obtained from PD, an example of PD Input file is present in
								<em>file_input_raw/Input_filter.xlsx</em>, the <strong>Sample</strong>
								column in optional, if is not present the software extract the names for
								the <strong>“File Name”</strong> column.
							</li>
							<li>In case of <strong>MQ</strong> analysis the this column is
								<code>REQUIRED</code>
							</li>
						</ul>
					</li>
				</ul>
			</li>
			<li>
				<p><code>Peptides file</code>: raw file of peptides obtained from PD
					or MQ (file <strong>peptides.txt</strong>).
				</p>
			</li>
			<li>
				<p><code>Proteins file</code>: raw file of protein groups obtained
					from PD or MQ (file <strong>proteinGroups.txt</strong>).
				</p>
			</li>
			<li>
				<p><code>Signal log2 expr thr</code>: signal log2 expression
					threshold for the differential analysis. <em>DEFAULT:</em> DEPs if
					<code>Signal log2 expression threshold &gt; -∞</code> (No Limit,
					represent by value “inf” in the cell)
				</p>
			</li>
			<li>
				<p><code>Log2 FC thr</code>: Fold Change threshold for the
					differential analysis. <em>DEFAULT:</em> DEPs if
					<code>log2 Fold Change threshold = 0.75</code> (Up-regulated &gt; 0.75,
					Down-regulated &lt; -0.75)
				</p>
			</li>
			<li>
				<p><code>P.Value thr</code>: p.value threshold for the differential
					analysis. <em>DEFAULT:</em> DEPs if
					<code>P value threshold &lt; 0.05</code>
				</p>
			</li>
			<li>
				<p><code>Batch Correction</code>: execution of the batch effect
					correction performed by <a href="https://www.bioconductor.org/packages/release/bioc/html/proBatch.html">proBatch</a>.
					<span class="citation">(Cuklina et al. 2018)</span>.
					(<strong>TRUE</strong> or <strong>FALSE</strong>). If
					<strong>TRUE</strong>, column <strong>MS_batch</strong> required in
					<strong>Input</strong> file.
				</p>
			</li>
			<li>
				<p><code>Control Boxplot proteins</code>: list of proteins used as
					control of the intensities. For each protein a boxplot is generated
					comparing the mean of the intensities group by condition.
				</p>
			</li>
			<li>
				<p><code>Design for the comparison file</code>: Excel file
					containing the formulas of the contrast comparison you want to analyse.
					The table can have 3 columns <button id="preview_design" type="button" class="btn_example" data-toggle="modal" data-target="#modal_Design_Example">[Example of Design file HERE...]</button>:
				</p>
				<ul>
					<li><strong>Formule</strong> column (<code>REQUIRED</code>): The
						formulas need to follow the syntax of <a href="https://bioinf.wehi.edu.au/limma/">Limma</a>.
						<code>AT LEAST 1 FORMULA IS REQUIRED</code>.
					</li>
					<li><strong>Name</strong> column (<code>OPTIONAL</code>): personalized
						name assign to the comparison.
					</li>
					<li><strong>Color</strong> columns (<code>OPTIONAL</code>): define a
						color for the condition in the graphs. If not present use a default
						palette.
					</li>
				</ul>
			</li>
			<li>
				<p><code>Execute Enrichment</code>: boolean value for the execution
					of the enrichment step. (<strong>TRUE</strong> or
					<strong>FALSE</strong>)
				</p>
			</li>
			<li>
				<p><code>Execute PPI network STRINGdb</code>: boolean value for the
					execution of the network analysis. (<strong>TRUE</strong> or
					<strong>FALSE</strong>)
				</p>
			</li>
			<li>
				<p><code>P.Value thr for enrichment</code>: p.value threshold of the
					enriched terms. <strong>DEFAULT:</strong> Term is significant if
					<code>P Value threshold for Enrichment &lt; 0.05</code>
				</p>
			</li>
			<li>
				<p><code>Overlap size thr for enrichment</code>: Overlap size
					threshold. The overlap size is the number of DEPs discovered in the
					enriched terms. <strong>DEFAULT:</strong> Term is significant if
					<code>Minimum number of overlap genes with enriched terms &gt; 5</code>
				</p>
			</li>
			<li>
				<p><code>Enrichment</code>: determine of filter the plot of the
					enrichment results. The enrichment can be filter by terms or by DBs. In
					the column <em>“Terms to search”</em> you can write the word that you
					want to search in the results of EnrichR (EX: MYC, C-MYC, Senescence,…).
					Instead, in the column <em>“DB to analyse”</em> you can write the DBs
					that you want to see in your plots (EX: ChEA_2016, KEGG_2021_Human,
					BioPlanet_2019, GO_Biological_Process_2021,…), using the same name that
					you can find in EnrichR. If the two columns are empty, no plots of
					enrichment are returned.
				</p>
			</li>
		</ul>
		</div>
    </li>
  </ul>
</div>
<div id="example-case-study---proteomics-of-mcf7-cells" class="section level2">
  <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
        	<h2>3. Example case study - Proteomics of MCF7 cells</h2>
        	<hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	      <p>Proteomics of MCF7 cells (LC-MS/MS Orbitrap Fusion Tribrid). PRIDE:
		<a href="http://proteomecentral.proteomexchange.org/cgi/GetDataset?ID=PXD009417">PXD009417</a>.
		Its execution is performed clicking the button <strong>Case Study
		Example</strong>. Clamer M, Tebaldi T, Lauria F, et al. Active Ribosome
		Profiling with RiboLace. Cell Rep. 2018;25(4):1097-1108.e5.
	</p>
      </div>
    </li>
  </ul>
</div>
<div id="details-on-the-output-files" class="section level2">
  <ul id="collapse_ul"  class="collapsible">
    <li id="collapse_li">
    	<div class="collapsible-header">
    	  <div style="width: 95%;">
        	<h2>4. Details on the output files</h2>
  	      <hr id="linesubtitle">
    	  </div>
    	  <i id="plus_elem" class="fa fa-plus"></i>
    	</div>
      <div class="collapsible-body">
	      <ul>
		<li><strong>report.html</strong>: complete report of the analysis with
			all pics and results of the enrichment.
		</li>
		<li><strong>database_env_R.RData</strong>: RData containing the
			essential variables for further analysis.
		</li>
		<li><strong>normalised_intensity_table.xlsx</strong>: excel file with
			the matrix used in the analysis. In sheet <em>Normalized proteins matrix
			intensity</em> is written the normalized and imputed version of the
			original intensities of each protein in each sample. In sheet
			<em>Normalized peptides matrix intensity</em> is written the normalized
			and imputed version of the original intensities of each peptide in each
			sample. In sheet <em>Normalized proteins intensity group by</em> is
			written the average and the standard deviation of each condition
			described in the Input file for each protein. In sheet <em>Normalized
			peptides intensity group by</em> is written the average and the standard
			deviation of each condition described in the Input file for each
			peptide.
		</li>
		<li>
			<strong>DEPs_table.xlsx</strong>: summary of the result of Limma. In
			sheet <em>DEPs_proteins</em>, Column <em>id</em> is the name of the
			protein. In sheet <em>DEPs_peptides</em>, column <em>Accession</em> is
			the UniprotID of protein, column <em>Description</em> is the description
			of protein, <em>GeneName</em> is Gene Symbol, <em>Annotated
			Sequence</em> is the sequence of peptides, <em>Modifications</em> is the
			modification that affect the peptides, <em>Position in Master
			Proteins</em> is the UniprotID with the position on the sequence. For
			each contrast rule are present 5 columns:
			<ul>
				<li><em>class</em>: is “+” if the protein is up-regulated following the
					parameters set in <strong>options.xlsx</strong>, “-” if the protein is
					down-regulated, “=” otherwise
				</li>
				<li><em>log2_FC</em>: is the log2 Fold Change of the contrast</li>
				<li><em>p_val</em>: is the p. value</li>
				<li><em>p_adj</em>: is the adjusted p. value</li>
				<li><em>log2_expr</em>: is the signal log2 expression</li>
			</ul>
		</li>
		<li><strong>general_mds_proteins.pdf</strong>: MDS of the samples by
			proteins.
		</li>
		<li><strong>general_mds_peptides.pdf</strong>: MDS of the samples by
			peptides.
		</li>
		<li><strong>corrected_mds_proteins.pdf</strong>: MDS of the samples by
			proteins after the bacth effect correction.
		</li>
		<li><strong>corrected_mds_peptides.pdf</strong>: MDS of the samples by
			peptides after the bacth effect correction.
		</li>
		<li><strong>boxplot_control_intensity.pdf</strong>: boxplot for the
			control of the intensities of selected proteins.
		</li>
		<li><strong>DEPs_proteins_count_barplot.pdf</strong>: number of DEPs
			find for each contrast design based on proteins. Light grey are the
			down-regulated, dark grey are the up-regulated genes.
		</li>
		<li><strong>DEPs_peptides_count_barplot.pdf</strong>: number of DEPs
			find for each contrast design based on peptides. Light grey are the
			down-regulated, dark grey are the up-regulated genes.
		</li>
		<li><strong>DEPs_proteins_mds.pdf</strong>: MDS of the DEPs discovered
			based on proteins.
		</li>
		<li><strong>DEPs_peptides_mds.pdf</strong>: MDS of the DEPs discovered
			based on peptides.
		</li>
		<li><strong>DEPs_enrichment_data.RData</strong>: result of the
			enrichment made with EnrichR.
		</li>
		<li><strong>DEPs_enrichment_table.xlsx</strong>: excel with only the
			enriched terms with an high significance (P.Value &lt; 0.05, Overlap
			Size &gt;= 4).
		</li>
		<li>
			<strong>enrich</strong> directory: plots of the enrichment data
			under the parameters described in the <strong>option.xlsx</strong> file
			<ul>
				<li><strong>DBs_enr.pdf</strong>: dot matrix of the enrichment result
					divided in up- and down-regulated genes. Filtered by the datasets choose
					in the <strong>options</strong> file.
				</li>
				<li><strong>terms_enr.pdf</strong>: dot matrix of the enrichment result
					divided in up- and down-regulated genes. Filtered by the terms written
					in the <strong>options</strong> file.
				</li>
				<li><strong>DBs_enr_all.pdf</strong>: dot matrix of the enrichment
					result combining the up- and down- regulated genes in a single list.
					Filtered by the datasets choose in the <strong>options</strong>
					file.
				</li>
				<li><strong>terms_enr_all.pdf</strong>: dot matrix of the enrichment
					result combining the up- and down- regulated genes in a single list.
					Filtered by the terms written in the <strong>options</strong> file.
				</li>
			</ul>
		</li>
		<li>
			<strong>network</strong> directory: plots of the up- and down-
			regulated genes based on STRINGdb. For each contrast design are create 2
			files:
			<!-- - **Prefilter**: histogram representing the number of genes in each community discovered. -->
			<ul>
				<li><strong>Postfilter</strong>: histogram representing the number of
					genes in each community discovered after a filter on the genes and on
					the links.
				</li>
				<li><strong>Graph</strong>: visual representation of the network
					discovered in two layouts: <em>fr</em> and <em>kk</em>.
				</li>
			</ul>
		</li>
	</ul>
      </div>
    </li>
  </ul>
</div>
<hr id="linesubtitle">
<div id="refs" class="references csl-bib-body hanging-indent">
		<div id="ref-igraph" class="csl-entry">
			Csardi, Gabor, and Tamas Nepusz. 2006. <span>“The Igraph Software
			Package for Complex Network Research”</span> Complex Systems: 1695. <a href="https://igraph.org">https://igraph.org</a>.
		</div>
		<div id="ref-proBatch" class="csl-entry">
			Cuklina, Jelena, Chloe H. Lee, Evan G. Willams, Ben Collins, Tatjana
			Sajic, Patrick Pedrioli, Maria Rodriguez-Martinez, and Ruedi Aebersold.
			2018. <span>“Computational Challenges in Biomarker Discovery from
			High-Throughput Proteomic Data.”</span> <a href="https://doi.org/10.3929/ethz-b-000307772">https://doi.org/10.3929/ethz-b-000307772</a>.
		</div>
		<div id="ref-STRINGdb" class="csl-entry">
			Szklarczyk, Damian, Annika L Gable, Katerina C Nastou, David Lyon,
			Rebecca Kirsch, Sampo Pyysalo, Nadezhda T Doncheva, et al. 2021.
			<span>“The STRING Database in 2021: Customizable Protein-Protein
			Networks, and Functional Characterization of User-Uploaded
			Gene/Measurement Sets.”</span> 49.
		</div>
	</div>
